Google.Api.Generator/Generation/MethodDetails.cs

```diff
@@ -253,37 +253,112 @@ private static MethodDetails DetectPagination(ServiceDetails svc, MethodDescript
+            }
+
+            // DiREGapic case where a return message has exactly one repeated field
+            // (pageSizeCandidate's name can be either "page_size" or "max_results")
+            if (repeatedCandidatesByDeclOrder.Count == 1 && !mapCandidates.Any())
+            {
+                return new Paginated(svc, desc, repeatedCandidatesByDeclOrder.Single(), pageSizeCandidate.FieldNumber, pageTokenCandidate.FieldNumber);
             }
-            return null;
 
-            bool IsCandidateField(FieldDescriptor field) =>
+            var errMsg = $"The method {desc.FullName} is selected as a pagination candidate " +
+                         $"but the configuration of the item response field candidates " +
+                         $"does not match any of the configurations we can generate.";
+                              
+            throw new InvalidOperationException(errMsg);
+
+            bool IsRepeatedCandidate(FieldDescriptor field) =>
                 field.IsRepeated && !field.IsMap && (field.FieldType == FieldType.Message || field.FieldType == FieldType.String);
         }
 
+        /// <summary>
+        /// Finds an appropriate candidate for a `page size`/`max results` request field.
+        /// That is a field that is named "page_size" or "max_results" and has a non-repeated Int32 type, if any.
+        /// If no such field exists, then simply a field that is named "page_size" or "max_results".
+        /// </summary>
+        private static FieldDescriptor FindPageSizeCandidate(MessageDescriptor input)
+        {
+            // Has the int32 page_size or int32 max_results field which defines the maximum number of paginated resources to return in the response
+            var pageSizeCandidate = input.FindFieldByName("page_size");
+            var maxResultsCandidate = input.FindFieldByName("max_results");
+
+            if (pageSizeCandidate is null && maxResultsCandidate is null)
+            {
+                return null;
+            }
+
+            if (IsValidPageSizeCandidate(pageSizeCandidate))
+            {
+                return pageSizeCandidate;
+            } 
+
+            if (IsValidPageSizeCandidate(maxResultsCandidate))
+            {
+                return maxResultsCandidate;
+            }
+
+            // returning the non-null candidate here even though it did not pass all the requirements
+            // an exception should be thrown but only if method has _all_ candidates, which we will only know later.
+            return pageSizeCandidate ?? maxResultsCandidate; 
+        }       
+
+        /// <summary>
+        /// The non-name criteria for a field to be a suitable candidate for a `page size`/`max results` request field.
+        /// </summary>
+        private static bool IsValidPageSizeCandidate(FieldDescriptor field) =>
+            field!=null && ((field.FieldType == FieldType.Int32 || field.FieldType == FieldType.UInt32) && !field.IsRepeated);
```

Comment:
Nit: `field!=null` => `field != null` (or `field is object`)

Reply:
👍 
also, my dissatisfaction about `is object` syntax, through the lens of C# documentation:
"The is operator checks if the result of an expression is compatible with a given type" (https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/is) 
why do we check that `FieldDescriptor` is compatible with `Object` again?
Oh, it's because the full specification clarifies that the null check is performed
"If E is a method group or the null literal, of if the type of E is a reference type or a nullable type and the value of E is null, the result is false."
(https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/expressions#the-is-operator)
at this point it's like the question of 'why are we hitting ourselves', luckily the C# 9 will give us `is not null` and make this moot.